File: app/src/main/java/net/ducksmanager/util/CoverFlowActivity.java
package net.ducksmanager.util;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageSwitcher;
import android.widget.ImageView;
import android.widget.TextSwitcher;
import android.widget.TextView;
import android.widget.Toast;

import net.ducksmanager.persistence.models.composite.CoverSearchIssueWithUserIssueDetails;
import net.ducksmanager.persistence.models.composite.InducksIssueWithUserIssueDetails;
import net.ducksmanager.whattheduck.AddIssue;
import net.ducksmanager.whattheduck.R;
import net.ducksmanager.whattheduck.WhatTheDuck;
import net.ducksmanager.whattheduck.WhatTheDuckApplication;

import java.util.List;

import androidx.appcompat.app.AppCompatActivity;
import it.moondroid.coverflow.components.ui.containers.FeatureCoverFlow;


public class CoverFlowActivity extends AppCompatActivity {

    private List<CoverSearchIssueWithUserIssueDetails> data;
    private TextSwitcher mResultNumber;
    private ImageSwitcher mCountryBadge;
    private ImageSwitcher mIssueCondition;

    private TextView mIssueConditionText;
    private TextView mTitleText;

    public static CoverSearchIssueWithUserIssueDetails currentSuggestion = null;

    CoverFlowAdapter adapter;
    FeatureCoverFlow coverFlow;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ((WhatTheDuckApplication) getApplication()).trackActivity(this);
        setContentView(R.layout.activity_coverflow);

        WhatTheDuck.appDB.coverSearchIssueDao().findAll().observe(this, searchIssues -> {
            data = searchIssues;
            adapter = new CoverFlowAdapter(this);
            adapter.setData(searchIssues);

            coverFlow = findViewById(R.id.coverflow);
            coverFlow.setAdapter(adapter);

            coverFlow.setOnItemClickListener((parent, view, position, id) -> {
                if (currentSuggestion.getUserIssue() == null) {
                    WhatTheDuck.setSelectedCountry (currentSuggestion.getCoverSearchIssue().getCoverCountryCode());
                    WhatTheDuck.setSelectedPublication (currentSuggestion.getCoverSearchIssue().getCoverPublicationCode());
                    WhatTheDuck.setSelectedIssue(currentSuggestion.getCoverSearchIssue().getCoverIssueNumber());

                    CoverFlowActivity.this.startActivity(new Intent(CoverFlowActivity.this, AddIssue.class));
                }
                else {
                    Toast.makeText(
                            CoverFlowActivity.this,
                            R.string.issue_already_possessed,
                            Toast.LENGTH_SHORT)
                            .show();
                }
            });

            coverFlow.setOnScrollPositionListener(new FeatureCoverFlow.OnScrollPositionListener() {
                @Override
                public void onScrolledToPosition(int position) {
                    currentSuggestion = data.get(position);

                    String uri = "@drawable/flags_" + currentSuggestion.getCoverSearchIssue().getCoverCountryCode();
                    int imageResource = getResources().getIdentifier(uri, null, getPackageName());

                    if (imageResource == 0) {
                        imageResource = R.drawable.flags_unknown;
                    }
                    mCountryBadge.setVisibility(View.VISIBLE);
                    mCountryBadge.setImageResource(imageResource);

                    mIssueConditionText.setVisibility(View.VISIBLE);

                    if (currentSuggestion.getUserIssue() != null) {
                        String condition = currentSuggestion.getUserIssue().getCondition();
                        mIssueCondition.setVisibility(View.VISIBLE);
                        mIssueCondition.setImageResource(InducksIssueWithUserIssueDetails.issueConditionToResourceId(condition));
                        mIssueConditionText.setText(InducksIssueWithUserIssueDetails.issueConditionToStringId(condition));
                        mIssueConditionText.setTextSize(18);
                    }
                    else {
                        mIssueCondition.setVisibility(View.GONE);
                        mIssueConditionText.setText(R.string.add_cover);
                        mIssueConditionText.setTextSize(14);
                    }

                    mResultNumber.setVisibility(View.VISIBLE);
                    mResultNumber.setText(getResources().getString(R.string.result) + " " + (position + 1) + "/" + data.size());

                    mTitleText.setVisibility(View.VISIBLE);
                    mTitleText.setText(data.get(position).getCoverSearchIssue().getCoverPublicationTitle() + " " + data.get(position).getCoverSearchIssue().getCoverIssueNumber());
                }

                @Override
                public void onScrolling() {
                    mResultNumber.setVisibility(View.INVISIBLE);
                    mTitleText.setVisibility(View.INVISIBLE);
                    mCountryBadge.setVisibility(View.INVISIBLE);
                    mIssueCondition.setVisibility(View.INVISIBLE);
                    mIssueConditionText.setVisibility(View.INVISIBLE);
                }
            });
        });

        mResultNumber = findViewById(R.id.resultNumber);
        mResultNumber.setFactory(() -> {
            LayoutInflater inflater = LayoutInflater.from(CoverFlowActivity.this);
            return inflater.inflate(R.layout.item_title, null);
        });

        mCountryBadge = findViewById(R.id.imageSwitcherCountryBadge);
        mCountryBadge.setFactory(() ->
            new ImageView(getApplicationContext()));

        mIssueCondition = findViewById(R.id.prefiximage);
        mIssueCondition.setFactory(() ->
            new ImageView(getApplicationContext()));

        TextSwitcher mIssueConditionTextSwitcher = findViewById(R.id.prefiximage_description);
        mIssueConditionTextSwitcher.setFactory(() -> {
            LayoutInflater inflater = LayoutInflater.from(CoverFlowActivity.this);
            mIssueConditionText = (TextView) inflater.inflate(R.layout.item_title, null);
            return mIssueConditionText;
        });

        TextSwitcher mTitleSwitcher = findViewById(R.id.title);
        mTitleSwitcher.setFactory(() -> {
            LayoutInflater inflater = LayoutInflater.from(CoverFlowActivity.this);
            mTitleText = (TextView) inflater.inflate(R.layout.item_title, null);
            return mTitleText;
        });
    }
}

File: app/src/main/java/net/ducksmanager/util/CoverFlowActivity.kt
package net.ducksmanager.util

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.AdapterView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import it.moondroid.coverflow.components.ui.containers.FeatureCoverFlow.OnScrollPositionListener
import net.ducksmanager.activity.AddIssues
import net.ducksmanager.persistence.models.composite.CoverSearchIssueWithDetails
import net.ducksmanager.persistence.models.composite.InducksIssueWithUserIssueAndScore.Companion.issueConditionToResourceId
import net.ducksmanager.persistence.models.composite.InducksIssueWithUserIssueAndScore.Companion.issueConditionToStringId
import net.ducksmanager.whattheduck.R
import net.ducksmanager.whattheduck.WhatTheDuck
import net.ducksmanager.whattheduck.WhatTheDuck.Companion.appDB
import net.ducksmanager.whattheduck.WhatTheDuck.Companion.selectedCountry
import net.ducksmanager.whattheduck.WhatTheDuck.Companion.selectedIssues
import net.ducksmanager.whattheduck.WhatTheDuck.Companion.selectedPublication
import net.ducksmanager.whattheduck.databinding.ActivityCoverflowBinding

class CoverFlowActivity : AppCompatActivity() {
    private lateinit var data: List<CoverSearchIssueWithDetails>
    private lateinit var binding: ActivityCoverflowBinding

    private lateinit var adapter: CoverFlowAdapter

    companion object {
        var currentSuggestion: CoverSearchIssueWithDetails? = null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (application as WhatTheDuck).trackActivity(this)
        binding = ActivityCoverflowBinding.inflate(layoutInflater)
        setContentView(binding.root)

        appDB!!.coverSearchIssueDao().findAll().observe(this, { searchIssues: List<CoverSearchIssueWithDetails> ->
            if (searchIssues.isEmpty()) {
                return@observe
            }
            data = searchIssues
            adapter = CoverFlowAdapter(this)
            adapter.setData(searchIssues)

            binding.coverflow.visibility = VISIBLE
            binding.coverflow.adapter = adapter
            binding.coverflow.onItemClickListener = AdapterView.OnItemClickListener { _: AdapterView<*>?, _: View?, _: Int, _: Long ->
                if (currentSuggestion!!.userIssue == null) {
                    appDB!!.inducksCountryDao().findByCountryCode(currentSuggestion!!.coverSearchIssue.coverCountryCode).observe(this, { country ->
                        selectedCountry = country
                        appDB!!.inducksPublicationDao().findByPublicationCode(currentSuggestion!!.coverSearchIssue.coverPublicationCode).observe(this, { publication ->
                            selectedPublication = publication
                            selectedIssues = mutableListOf(currentSuggestion!!.coverSearchIssue.coverIssueNumber)
                            this@CoverFlowActivity.startActivity(Intent(this@CoverFlowActivity, AddIssues::class.java))
                        })
                    })
                } else {
                    Toast.makeText(
                        this@CoverFlowActivity,
                        R.string.issue_already_possessed,
                        Toast.LENGTH_SHORT)
                        .show()
                }
            }
            binding.coverflow.setOnScrollPositionListener(object : OnScrollPositionListener {
                override fun onScrolledToPosition(position: Int) {
                    currentSuggestion = data[position]
                    val uri = "@drawable/flags_" + currentSuggestion!!.coverSearchIssue.coverCountryCode
                    var imageResource = resources.getIdentifier(uri, null, packageName)
                    if (imageResource == 0) {
                        imageResource = R.drawable.flags_unknown
                    }

                    toggleInfoVisibility(VISIBLE)

                    binding.countrybadge.setImageResource(imageResource)

                    val condition = currentSuggestion!!.userIssue?.condition
                    if (condition != null) {
                        val conditionResourceId = issueConditionToResourceId(condition)
                        if (conditionResourceId != null) {
                            binding.conditionbadge.setImageResource(conditionResourceId)
                            binding.conditiontext.text = getString(issueConditionToStringId(condition))
                        }
                        binding.conditiontext.textSize = 18f
                    } else {
                        binding.conditionbadge.visibility = View.GONE
                        binding.conditiontext.text = getString(R.string.add_cover)
                        binding.conditiontext.textSize = 14f
                    }

                    if (currentSuggestion!!.suggestionScore <= 0) {
                        binding.score.root.visibility = View.GONE
                    }
                    binding.score.scorevalue.text = currentSuggestion!!.suggestionScore.toString()
                    binding.score.scorevalue.textSize = 20F

                    val quotation = currentSuggestion!!.coverSearchIssue.quotation
                    if (quotation != null) {
                        binding.quotation.visibility = VISIBLE
                        if (quotation.containsKey("min") && quotation.containsKey("max")) {
                            binding.quotationvalue.text = String.format(resources.getString(R.string.quotation_between), quotation["min"], quotation["max"])
                        }
                        else if (quotation.containsKey("min")) {
                            binding.quotationvalue.text = String.format(resources.getString(R.string.quotation_more_than), quotation["min"])
                        }
                        else {
                            binding.quotationvalue.text = String.format(resources.getString(R.string.quotation_less_than), quotation["max"])
                        }
                    }
                    else {
                        binding.quotation.visibility = GONE
                    }

                    binding.resultNumber.text = resources.getString(
                        R.string.coversearch_result_template,
                        resources.getString(R.string.result),
                        position + 1,
                        data.size
                    )

                    binding.issuetitle.text = resources.getString(
                        R.string.title_template,
                        data[position].coverSearchIssue.coverPublicationTitle,
                        data[position].coverSearchIssue.coverIssueNumber
                    )
                }

                override fun onScrolling() {
                    toggleInfoVisibility(View.INVISIBLE)
                }

                fun toggleInfoVisibility(visibility: Int) {
                    listOf(
                        binding.countrybadge,
                        binding.issuetitle,
                        binding.conditionbadge,
                        binding.conditiontext,
                        binding.score.root,
                        binding.quotation,
                        binding.resultNumber
                    ).forEach{ it.visibility = visibility }
                }
            })
        })
    }
}